content/blog/2025-12-16-super-vectorization.md

```diff
@@ -0,0 +1,100 @@
+
+```pseudocode
+// Original straight-line code
+x1 = a1 + b1
+x2 = a2 + b2
+y1 = c1 * d1
+y2 = c2 * d2
+
+// Grouped for vectorization
+vec_x = vec_add(vec_a, vec_b)
+vec_y = vec_mul(vec_c, vec_d)
+```
+
+## The Paper and my Goal
+
+### The Paper's Approach
+
+In ["All You Need Is Superword-Level Parallelism,"](https://dspace.mit.edu/bitstream/handle/1721.1/146343/3519939.3523701.pdf?isAllowed=y&sequence=1) Chen et. al. outline a simple principle: if the compiler can reorder instructions freely and easily determine which instructions can be packed together, the same SLP algorithm which only works in basic blocks suddenly can generalize across a whole function (or, with inlining, a who program). But, LLVM's native CFG based IR does not allow this: moving instructions is inherently tricky, and can easily result in the whole program becoming invalid.
+
+The authors solve this by introducing **Predicated SSA**, which when combined with an SLP algorithm they call **Super Vectorization**. Predicated SSA solves the issue of a CFG via removing the CFG. Instead, a function is one list of straight line code, composed entirely of loops and instructions. To ensure data and control dependencies remain, each instruction or loop has an associated predicate: the code only runs if the predicate is satisfied. Thus, while data dependencies still have to be respected, control dependencies are made irrelevant, turned into a simply equality check
+
+**Traditional CFG:**
+```
+Start -> If cond -> Then Block -> End
+         -> Else Block -> End
+```
+
+**Predicated SSA:**
+```
+pred1 = cond
+instr1 (pred1)  // runs if cond
+instr2 (!pred1) // runs if !cond
+```
+
+### My Goal
+
+My goal was to implement some form of Super Vectorization, however upon further inspection the paper, while at a fundamental level about vectorization, was more about the form of the IR the paper presented. The authors took almost eight-thousand lines to implement the full pipeline, therefore I limited myself to implementing the IR and pieces of the optimizations. For the IR, a full implementation was the set goal, letting any program in the LLVM IR be translated into Predicated SSA. On optimizations, this was kept fairly minimal, as the point of the project was more the framework than the optimizations themselves. Basic vector packing was achieved, following data dependencies and straight line code.
+
+## Implementation
+
+Source Code: [HERE](https://github.com/tf-mac/superVectorization)
+
+The overall implemention took about a thousand lines of C++ code. Most of that was spent on the conversion to and from Predicated SSA, as the intention of the IR is to make any post-translation analysis relatively simple. My process for implementing the IR was to outline the semantics, then build up skeletons of both functions and finally fill in the relevant helper functions. Both functions had some difficulty both from learning how to implement the ideas and from the paper itself.
```

Comment:
Can you give us any kind of a glimpse into what your IR data structure looks like? Here are some questions that naturally arise:

* Is every function just a list of (predicated) instructions? Or is it more complicated than that?
* Do those instructions refer to LLVM `Instruction` objects, or is it completely independent?
* How are predicates represented? Can they be an arbitrary program variable? Or are there special predicate expression trees?